### PR TITLE
Add community detection

### PR DESCRIPTION
**Pull request by rahuljoglekar47**
## Fixes # .
Code should be reasonably efficient (hasn't been profiled yet) as a result of restructuring it; the Python logging module was also implemented

## Summary/Motivation:
Take in a Pyomo optimization model into function detect_communities, and then return a map of the communities within the model (based on various user-specified parameters such as the node type, whether the graph is weighted, etc.)

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by rahuljoglekar47** [comment]
Please let me know if there is a better way to determine the variables in a given constraint equation (as I did in lines 57 and 58 of community_graph.py); I do something like this (adding the string of a variable to a list) many times in this file so I was wondering if this is the best way to create a list of variables in a given equation (or constraints that contain a given variable).

File: pyomo/contrib/community_detection/detection.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
+"""Community Detection Code - Rahul Joglekar"""
```
Comment (bernalde) [inline pyomo/contrib/community_detection/detection.py]: Usually contributed files to Pyomo do not include the author.

File: pyomo/contrib/community_detection/detection.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+    if number_of_communities == 0:
+        logging.error("in detect_communities: Empty community map was returned")
+    if number_of_communities == 1:
+        logging.warning("Community detection found that with the given parameters, the model could not be decomposed - "
+                        "only one community was found")
+
+    return community_map
+
+
+def check_for_correct_arguments(model, node_type, with_objective, weighted_graph, file_destination, log_level,
```
Comment (bernalde) [inline pyomo/contrib/community_detection/detection.py]: Since this is some sort of internal function with a generic name I would suggest renaming it as `_check_for_correct_arguments`

File: pyomo/contrib/community_detection/tests/test_detection.py [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+from pyomo.contrib.community_detection.detection import *
+
+from pyomo.solvers.tests.models.LP_unbounded import LP_unbounded
+from pyomo.solvers.tests.models.QP_simple import QP_simple
+from pyomo.solvers.tests.models.LP_inactive_index import LP_inactive_index
+from pyomo.solvers.tests.models.SOS1_simple import SOS1_simple
+
+
+class TestDecomposition(unittest.TestCase):
+
```
Comment (bernalde) [inline pyomo/contrib/community_detection/tests/test_detection.py:35]: define a variable `random_seed=5` here and use it below instead of hardcoding the 5 in every instance

File: pyomo/contrib/community_detection/tests/test_detection.py [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+    #                   x        b        i      s1s      s2s       sc       si
+    #      Total     cont   binary  integer     sos1     sos2    scont     sint
+    #          7        1        0        6        0        0        0        0
+    #  FX      0        0        0        0        0        0        0        0
+    #
+    #  Nonzero counts
+    #      Total    const       NL      DLL
+    #         37       35        2        0
+    #
+    #  Reformulation has removed 1 variable and 1 equation
```
Comment (bernalde) [inline pyomo/contrib/community_detection/tests/test_detection.py]: The comments here and above can be removed. You can add a single line saying that it comes from instance XXX.gms in minlplib.com

File: pyomo/contrib/community_detection/tests/test_detection.py [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+    model = m = ConcreteModel()
+    m.i1 = Var(within=Integers, bounds=(0, 1E15), initialize=0)
+    m.i2 = Var(within=Integers, bounds=(0, 1E15), initialize=0)
+    m.i3 = Var(within=Integers, bounds=(0, 1E15), initialize=0)
```
Comment (bernalde) [inline pyomo/contrib/community_detection/tests/test_detection.py]: The upper bounds of integer variables being this large seem wrong. Just set 100 (GAMS does that by default).

File: pyomo/contrib/pynumero/examples/structured/tests/test_nlp_transformations.py [hunk 6]
```diff
@@ -12,7 +12,8 @@
 import pyomo.environ as aml
 import os
 
-from .. import numpy_available, scipy_available
+#from .. import numpy_available, scipy_available
```
Comment (bernalde) [inline pyomo/contrib/pynumero/examples/structured/tests/test_nlp_transformations.py]: This shouldn't be here. These changes are not part of this PR, so revert them to original state.

File: pyomo/solvers/tests/models/LP_unique_duals.py [hunk 7]
```diff
@@ -8,7 +8,8 @@
 #  This software is distributed under the 3-clause BSD License.
 #  ___________________________________________________________________________
 
-from pyomo.core import ConcreteModel, Param, Var, Expression, Objective, Constraint, RangeSet, ConstraintList, NonNegativeReals, Suffix, summation
+from pyomo.core import ConcreteModel, Param, Var, Expression, Objective, Constraint, RangeSet, ConstraintList, \
+    NonNegativeReals, Suffix, summation, sum_product
```
Comment (bernalde) [inline pyomo/solvers/tests/models/LP_unique_duals.py]: Same case here, do not edit any of the files that do not belong to the original contribution.
Were these changes required for the tests of communities to work?

File: pyomo/contrib/community_detection/detection.py [hunk 8]
```diff
@@ -0,0 +1,156 @@
+"""Community Detection Code - Rahul Joglekar"""
+from pyomo.core import ConcreteModel
+from contrib.community_detection import community_graph
+import community
```
Comment (bernalde) [inline pyomo/contrib/community_detection/detection.py]: This is a point that we haven't discussed yet but that seems important. What happens if the user does not have the package `community` installed? We should have a way of checking this import and besides hopefully having this package installed along with Pyomo.
Follow the code here https://github.com/Pyomo/pyomo/blob/07d3cdbde1139d3f926d322132c1995e00187b62/pyomo/contrib/mindtpy/mip_solve.py#L28 to attempt an import of the package and add the documentation of the package in the `communities.rst` file

**Review by bernalde — CHANGES_REQUESTED** [review]
Hi Rahul, this looks fine and it should be working.
My two main requests are:

1. Adding an attempt import for the community package. If the user does not have it installed, then you should also skip the tests (at least those tests that require you running the community detection algorithm).
2. Adding the documentation of your contributed package as an .rst file
